### PR TITLE
Allow Decimal to be used in ipmt, pmt, ppmt, nper, and fv

### PR DESCRIPTION
**Pull request by garrypolley**
This change allows for a custom type to be passed for each of the values in ipmt, pmt, and ppmt. 

Example:

```py
import numpy as np
from decimal import Decimal

rate = np.array(Decimal('10'), dtype=Decimal)
nper = np.array(Decimal('1'), dtype=Decimal)
per = np.array(Decimal('1'), dtype=Decimal)
pv = np.array(Decimal('25000.00'), dtype=Decimal)

np.ipmt(rate/Decimal(360), per, nper, pv, fv=Decimal(0.0), when=0)
>>> array(Decimal('-694.4444444444444444444444445'), dtype=object)

np.ipmt(Decimal(10)/Decimal(360), Decimal(1), Decimal(1), Decimal(25000), Decimal(0.0))
>>> array(Decimal('-694.4444444444444444444444445'), dtype=object)

np.pmt(rate/Decimal(360), nper, pv, fv=Decimal(0.0))
>>> Decimal('-25694.44444444444444444444424')

np.pmt(Decimal(10)/Decimal(360), Decimal(1), Decimal(25000), fv=Decimal(0.0))
>>> Decimal('-25694.44444444444444444444424')
```

For fun you can see how float is still different:

```python
np.pmt(10/360, 1, 25000)
>>>  -25694.444444444533
```

If this is a desired/wanted feature I can add some tests specifically for passing `Decimal` as the type. 

Note: if all the tests fail in terrible ways because of this change, I'll likely close this PR in shame. 😸

Also, I'll update my commit message to match the guidelines if this feature enhancement is desired.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Just in case you though it was doing something clever, `dtype=Decimal` is equivalent to `dtype=object`.

File: numpy/lib/financial.py [hunk 1]
```diff
@@ -116,11 +124,12 @@ def fv(rate, nper, pmt, pv, when='end'):
     """
     when = _convert_when(when)
     (rate, nper, pmt, pv, when) = map(np.asarray, [rate, nper, pmt, pv, when])
-    temp = (1+rate)**nper
+    one = _generate_typed_number(rate, 1)
+    temp = (one+rate)**nper
     miter = np.broadcast(rate, nper, pmt, pv, when)
-    zer = np.zeros(miter.shape)
+    zer = np.zeros(miter.shape, dtype=rate.dtype)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: This is overkill - just use `dtype=int`

**Review by eric-wieser — COMMENTED** [review]
In general, you've taken a more complex and fragile solution than you need to. Just replace all the floats with integers, and I think it will work :)

File: numpy/lib/financial.py [hunk 2]
```diff
@@ -209,11 +218,12 @@ def pmt(rate, nper, pv, fv=0, when='end'):
     when = _convert_when(when)
     (rate, nper, pv, fv, when) = map(np.array, [rate, nper, pv, fv, when])
     temp = (1 + rate)**nper
-    mask = (rate == 0.0)
-    masked_rate = np.where(mask, 1.0, rate)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Using `1` instead of `1.0` would be enough here

**Comment by eric-wieser** [comment]
> If this is a desired/wanted feature I can add some tests specifically for passing Decimal as the type.

If it's easy to support it, then I see no reason not to. Please do write these tests.

File: numpy/lib/financial.py (hunk 2 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I guess first pass with these changes going with integers should be fine.  The reason I was a bit more dynamic was an attempt to support other types, than only decimal. But first pass I should probably only do decimal.

Comment (eric-wieser) [reply numpy/lib/financial.py]: I figure that any type that knows how to add itself to integers is enough. I can't come up with any financially useful type that wouldn't allow that.

**Comment by garrypolley** [comment]
I plan to come back to this PR in about a week with the proposed changes and tests.

**Comment by garrypolley** [comment]
I have not forgotten about this code. I do still have plans to fix.  Sorry for the delay.

**Comment by garrypolley** [comment]
The work here is replaced by a newer/cleaner PR:  

https://github.com/numpy/numpy/pull/9952

**Pull request closed without merging**